### PR TITLE
Feat recursive verifier

### PR DESCRIPTION
**Pull request by enricobottazzi**
Issue #7 

Write up here => https://hackmd.io/bCCfhHRPS5S4zzNCpuWAIw

### CODE REVIEW

**Review by alxkzmn — APPROVED** [review]

File: zk_prover/src/circuits/tests.rs [hunk 1]
```diff
@@ -104,6 +113,131 @@ mod test {
         assert!(full_verifier(&params, &vk, proof, &public_input));
     }
 
+    #[test]
+    #[ignore]
+    fn test_valid_merkle_sum_tree_with_full_recursive_prover() {
+        // params for the aggregation circuit
+        let params_agg = generate_setup_params(23);
```
Comment (sifnoc) [inline zk_prover/src/circuits/tests.rs:120]: Could you provide some insight on why we need such a large number of rows here?

Comment (enricobottazzi) [reply zk_prover/src/circuits/tests.rs:120]: Sure, I'm quoting Han here =>

The k of aggregation circuit would be roughly `log2(30000*n)` where n is
`n = #columns + #internal_columns + #quotient_commitments + #pc_opening_commitments + 2*#old_accumulators`
where
- #columns are allocated by user e.g. meta.xxx_column() or meta.selector()
- #internal_columns are used by lookup argument and permutation argument
- #pc_opening_commitment depends on which kind of polynomial commitment scheme we are using

So in short it is mainly only about the #columns of the input circuit, so more columns -> larger aggregation circuit -> longer aggregation time. 

I think we can potentially move it down to 22, but not much more that this. As you can notice [here](https://github.com/privacy-scaling-explorations/snark-verifier/blob/main/snark-verifier-sdk/examples/standard_plonk.rs#L162) also the simple plonk example provided in the SDK requires k to be 22

**Review by sifnoc — APPROVED** [review]
Fantastic! This is another scalable solution for us!

File: zk_prover/src/circuits/tests.rs (hunk 1 above)
Comment (enricobottazzi) [reply zk_prover/src/circuits/tests.rs:120]: Hi @sifnoc, for `generate_setup_params` I suggest you downloading the relative `hermez_raw` file from https://github.com/han0110/halo2-kzg-srs. 

In the `README` it is described how to use that => https://github.com/summa-dev/summa-solvency/tree/master/zk_prover#powers-of-tau-trusted-setup

Also, in the latest optimization the `k` of the aggregation circuit will likely to be reduced to 21 or 20
